### ISSUE TITLE
State gets lost when the form is inside a conditional rendering

### ISSUE THREAD

**Issue by nacho-vazquez** [TanStack/form#567]
### Describe the bug

When updating the form value using either `form.setFieldValue` or 

```ts
form.store.setState((state) => {
        return {
          ...state,
          values: {
            ...data,
          },
        };
      });
 ```
 
 And having the form inside a conditional rendering like 
 
 ```ts
 if (xData) {
    return (
      <div>
        <form.Provider>
          <form>
            <form.Field
              name="firstName"
              validators={{ onBlur: z.string() }}
              children={(field) => (
                <input
                  id={field.name}
                  name={field.name}
                  value={field.state.value}
                  onBlur={field.handleBlur}
                  onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
                    field.handleChange(event.target.value)
                  }
                />
              )}
            ></form.Field>
          </form>
        </form.Provider>
      </div>
    );
  }
 ```
 
 The state gets lost, and it is never shown in the field.

### Your minimal, reproducible example

https://stackblitz.com/edit/stackblitz-starters-oj2quu?file=src%2FApp.tsx

### Steps to reproduce

1. Create a form
2. Update the values of the form asynchronously
3. Put the form fields inside a conditional rendering
4. See how the form input doesn't show the updated value.

### Expected behavior

As a user, I expected to see the updated values populate the fields.

### How often does this bug happen?

Every time

### Screenshots or Videos

https://github.com/TanStack/form/assets/149403662/c09878d5-7597-4719-ab97-91bb0fa9e229

### Platform

macOS - Chrome (Arc)

### Tanstack Form adapter

react-form

### TanStack Form version

0.13.3

### TypeScript version

5.2.2

**Comment by nacho-vazquez** [comment]
Sometimes the value flashes before getting cleared.

**Comment by nacho-vazquez** [comment]
Crazy enough, the problem seems to happen when the set state function `setxData(data);` is used

```ts
useEffect(() => {
    async function fetchData() {
      const p1 = new Promise<{ firstName: string }>((res) =>
        setTimeout(() => res({ firstName: 'Dave' }), 1000)
      );

      const data = await p1;

      setxData(data);

      form.store.setState((state) => {
        return {
          ...state,
          values: {
            ...data,
          },
        };
      });
    }

    fetchData();
  }, []);
  ```

**Comment by fulopkovacs** [comment]
# TL;DR
I looked into it and apparently when the state of the component in which the `useForm()` hook is updated, the component will be rerendered, which means that the `useForm()` hook runs once again and re-creates the form with the default values, ignoring the previous state of the form. 

The [code sandbox](https://stackblitz.com/edit/stackblitz-starters-oj2quu?file=src%2FApp.tsx) in the issue's description can be fixed, if the default values are set using `xData`:
```tsx
  const [xData, setxData] = useState(null);

  const form = useForm<{ firstName: string }>({
    defaultValues: {
      firstName: xData.firstName,
    },
    onSubmit: async ({ value }) => {
      console.log(value);
    },
  });
```


# Profiling the components (with React Dev Tools)

I compared how the same component (very similar to the one in the issue's description) changes when we update the component's state (`<App />`), vs when we don't.


## 1. Not modifying the state of `<App />` during data fetching
1 commit.

![image](https://github.com/TanStack/form/assets/43729152/9a8d2ed6-3e6a-4ff4-aa93-5b368e9732f3)
`Field` changes because one of its hooks. This must be because we changed its value when we called `form.setFieldValue('firstName', data.firstName)`.

## 2. Modifying the state of `<App />` during data fetching (second image)

3 commits.

### Commit 1
![image](https://github.com/TanStack/form/assets/43729152/636dae2d-829e-438a-8c88-dac3d4baa7aa)
 The `Field` rerenders, because one of its hooks has changed, I think this is when we update it through the api. I think this is the point when see the value we set in the Field.



### Commit 2
| `<App />` | `<Field />` |
| --- | --- |
| <img width="749" alt="image" src="https://github.com/TanStack/form/assets/43729152/20e02444-1b84-4111-8bb6-4cfdb4a805f3"> | ![image](https://github.com/TanStack/form/assets/43729152/acaa4ab2-8afe-40a3-9035-29e589dcae68) |

`<App />` rerenders, because one of its hooks has changed. I think this is the `setXData` hook. 



This causes  `<Field/>` to rerender too. This must be the point when our value disappears from the field and it starts with a blank state again.

### Commit 3
![image](https://github.com/TanStack/form/assets/43729152/e7f1d744-3a9e-4e49-81c9-679ec9bf5907)

 Nothing important happens. We can ignore it.

**Comment by fulopkovacs** [comment]
I could look into this issue even further (but unfortunately only after a few days), but I'm not that familiar with the internals of the form.

@crutchcorn can you give any pointers? My intuition is that `form.Provider` creates a new React context, in which case we might have a problem because I think that context is recreated when the component that contains `form.Provider` is rerendered. This kind of situation would be solvable by storing the state of the form outside of the component that contains it. We would have to use that data to set the initial values of the form when its parent component gets rerendered.

The first idea that comes to my mind for implementing this is a global form context that is at the root of the whole React app, but sounds like a pretty big change.

**Comment by nacho-vazquez** [comment]
@fulopkovacs, thanks for the research. This is great!

**Comment by crutchcorn** [comment]
Sorry for the long wait @nacho-vazquez!

There's been a few shifts since you last opened this issue, and there **was** a bug that was fixed, but your code sample _still_ doesn't work out of the box.

That said, I'm unfortunately going to mark this as "working as intended" albiet confusingly. Here's why:

See, you can also change:

```tsx
form.setFieldValue('firstName', 'Davex');
```

To:

```tsx
form.setFieldValue('firstName', 'Davex', { touch: true });
```

To solve the issue.

By default, `touch` is `false`, which allows you to set the `fieldValue` programmatically internally without marking it as a user input

But when the re-render occurs with a new `defaultValue`, it throws away the previous field value since it's not a user input.

It's a bit confusing, but I generally consider the `setX` APIs in TanStack Form as internals.

Instead, we're investigating a story to support async default values easier.

Closing this for now, as a result of above, but don't be a stranger!

**Comment by nacho-vazquez** [comment]
It makes sense. I can make this work; thanks for looking into it.

**Comment by devth** [comment]
I'm running into a similar issue where I use `field.setValue` to programmatically update my form state but the UI doesn't reflect it. I'm new to this project so I'm probably missing something obvious!

**Comment by Balastrong** [comment]
@devth you may be looking for [Reactivity](https://tanstack.com/form/latest/docs/framework/react/guides/basic-concepts#reactivity) to explicitly subscribe to value changes.

**Comment by devth** [comment]
Thanks! I did indeed get it working with `Subscribe`. I think my biggest surprise / misunderstanding was having access to `person` in a loop like this, but not having that value update automatically like when using React's `useState`:

```typescript
        <form.Field name="people" mode="array">
          {(field) => {
            return (
              <div>
                {field.state.value.map((person, i) => {
                  return (
                    <form.Field key={i} name={`people[${i}].name`}>
                      {(subField) => {
                        return (
                          <div>
                            <label>
                              <div>Name for person is {person.name} –– THIS won't update automatically!!</div>
                              <input
                                value={subField.state.value}
                                onChange={(e) =>
                                  subField.handleChange(e.target.value)
                                }
                              />
                            </label>
                          </div>
                        )
                      }}
                    </form.Field>
                  )
                })}
                <button
                  onClick={() => field.pushValue({ name: '', age: 0 })}
                  type="button"
                >
                  Add person
                </button>
              </div>
            )
          }}
        </form.Field>
```
